samza-core/src/main/java/org/apache/samza/task/RunLoop.java

```diff
@@ -457,17 +457,25 @@ private void process() {
       final IncomingMessageEnvelope envelope = state.fetchEnvelope();
       log.trace("Process ssp {} offset {}", envelope.getSystemStreamPartition(), envelope.getOffset());
 
-      final ReadableCoordinator coordinator = new ReadableCoordinator(task.taskName());
-      TaskCallbackFactory callbackFactory = new TaskCallbackFactory() {
-        @Override
-        public TaskCallback createCallback() {
-          state.startProcess();
-          containerMetrics.processes().inc();
-          return callbackManager.createCallback(task.taskName(), envelope, coordinator);
-        }
+      Runnable processWorker = () -> {
+        final ReadableCoordinator coordinator = new ReadableCoordinator(task.taskName());
+        TaskCallbackFactory callbackFactory = new TaskCallbackFactory() {
+          @Override
+          public TaskCallback createCallback() {
+            state.startProcess();
+            containerMetrics.processes().inc();
+            return callbackManager.createCallback(task.taskName(), envelope, coordinator);
+          }
+        };
+
+        task.process(envelope, coordinator, callbackFactory);
       };
 
-      task.process(envelope, coordinator, callbackFactory);
+      if (threadPool != null) {
```

Comment:
This is a bad pattern in code. Can we fix atleast this instance of it. 
This threadPool is being passed from SamzaContainer as a null. 

```val taskThreadPool = if (threadPoolSize > 0) {
      Executors.newFixedThreadPool(threadPoolSize,
        new ThreadFactoryBuilder().setNameFormat("Samza Container Thread-%d").build())
    } else {
      null
    }```
a. We can either push the thread-pool creation down to the place its actually being used. 
b. Alternatively, can create it as a single-threaded executor pool.
c. or pass it as an optional.

Reply:
I am all in for this. However, I'd prefer to defer it to a follow up PR to clean up the semantics of threadPool and update the RunLoop code all at once instead of fixing this place alone. 
In my opinion, it is going to add more noise to the PR and makes it hard to debug issues associated with the PR. 
what do you think?